### DISCUSSION TITLE
Labwc Application Switcher

### DISCUSSION THREAD

**Discussion by stefonarch** [labwc/labwc#1510] [Show and tell]
Playing with [lswt](https://sr.ht/~leon_plickat/lswt/), wlrctl and qarma and some help (works also with zenity but window size has to be adapted). Works better with keyboard navigation as with mouse (2 clicks).

![screen_area_mar_22:54:01_](https://github.com/labwc/labwc/assets/10681413/d4d09a42-5611-4979-b986-3aaa2740fb10)

```
#!/bin/bash

applist=$(lswt | sed -e '1d')
str="${applist[*]}"

IFS=$'\n' arr=( $(xargs -n1 <<<"$str") )

selected_app=$(zenity --title="Labwc App Switcher" --list --column="" --hide-header --column="" --hide-header --width=700 --height=300 --ok-label="Activate" "${arr[@]}")

if [ -n "$selected_app" ]; then
    exec wlrctl window focus "$selected_app"
fi
```

**Comment by Narrat** [comment]
Very interesting idea.
Although I only get an empty `arr` when I tried this script and zenity therefore complains about not supported input.

Tinkered with the script with my severly lacking shell scripting ability, but I got something "working" for me.
Why the airquotes? It will for example break in a subtle way if the active tab in firefox contains `'` or similar string related chars...
And `wlrctl` seems to be severly lacking in its ability to focus windows? Only accepts `app_id`? Not that practical if multiple instances of a program are open...

New dep: `jq`

```
#!/bin/bash

IFS=$'\n' arr=( $(lswt -j | jq -r '.[]|.app_id+"\n",.title+"\n"') )

selected_app=$(zenity --title="Labwc App Switcher" --list --column="" --hide-header --column="" --hide-header --width=700 --height=300 --ok-label="Activate" "${arr[@]}")

if [ -n "$selected_app" ]; then
    exec wlrctl window focus "$selected_app"
fi
```

Not that much of a `zenity` user myself therefore next step will be looking into combining it with `bemenu`  or other `dmenu`-like programs

**Reply by Consolatis** [reply]
> And `wlrctl` seems to be severly lacking in its ability to focus windows? Only accepts `app_id`? Not that practical if multiple instances of a program are open...

You should be able to use something like `wlrctl window focus "app_id:${selected_app}" "title:${the_title}"`

**Reply by Narrat** [reply]
Oh indeed, that works. Thank you for the info.
That alone opens room for improvements. (e.g. getting rid of the second column)

```
#!/bin/bash

IFS=$'\n' arr=( $(lswt -j | jq -r '.[].title+"\n"') )

selected_app=$(zenity --title="Labwc App Switcher" --list --column="" --hide-header --width=700 --height=300 --ok-label="Activate" "${arr[@]}")

if [ -n "$selected_app" ]; then
    exec wlrctl window focus title:"$selected_app"
fi
```

**Comment by stefonarch** [comment]
Anyway, this is a temp workaround until https://github.com/labwc/labwc/pull/1479 is merged. It's notable more slow than the native switcher.
